dialect.go

```diff
@@ -1,364 +1,41 @@
 package goose
 
 import (
-	"database/sql"
 	"fmt"
+
+	"github.com/pressly/goose/v3/internal/dialect"
 )
 
-// SQLDialect abstracts the details of specific SQL dialects
-// for goose's few SQL specific statements
-type SQLDialect interface {
-	createVersionTableSQL() string // sql string to create the db version table
-	insertVersionSQL() string      // sql string to insert the initial version table row
-	deleteVersionSQL() string      // sql string to delete version
-	migrationSQL() string          // sql string to retrieve migrations
-	dbVersionQuery(db *sql.DB) (*sql.Rows, error)
+func init() {
+	store, _ = dialect.NewStore(dialect.Postgres, TableName())
 }
 
-var dialect SQLDialect = &PostgresDialect{}
-
-// GetDialect gets the SQLDialect
-func GetDialect() SQLDialect {
-	return dialect
-}
+var store dialect.Store
 
-// SetDialect sets the SQLDialect
-func SetDialect(d string) error {
-	switch d {
+// SetDialect sets the dialect to use for the goose package.
+func SetDialect(s string) error {
+	var d dialect.Dialect
+	switch s {
 	case "postgres", "pgx":
-		dialect = &PostgresDialect{}
+		d = dialect.Postgres
 	case "mysql":
-		dialect = &MySQLDialect{}
+		d = dialect.Mysql
 	case "sqlite3", "sqlite":
-		dialect = &Sqlite3Dialect{}
+		d = dialect.Sqlite3
 	case "mssql":
-		dialect = &SqlServerDialect{}
+		d = dialect.Sqlserver
 	case "redshift":
-		dialect = &RedshiftDialect{}
+		d = dialect.Redshift
 	case "tidb":
-		dialect = &TiDBDialect{}
+		d = dialect.Tidb
 	case "clickhouse":
-		dialect = &ClickHouseDialect{}
+		d = dialect.Clickhouse
 	case "vertica":
-		dialect = &VerticaDialect{}
+		d = dialect.Vertica
 	default:
-		return fmt.Errorf("%q: unknown dialect", d)
-	}
-
-	return nil
-}
-
-////////////////////////////
-// Postgres
```

Comment:
All these queries were moved into the `internal/dialect/dialectquery` package and exposed as a single interface:

```go
type Querier interface {
	CreateTable() string
	InsertVersion() string
	DeleteVersion() string
	GetMigrationByVersion() string
	ListMigrations() string
}
```

This makes it really trivial to have a `Store` interface that calls into this interface that depends on the underlying dialect, e.g.

```go
func (s *store) CreateVersionTable(ctx context.Context, tx *sql.Tx) error {
	q := s.querier.CreateTable()
	_, err := tx.ExecContext(ctx, q)
	return err
}
```